src/Symfony/Component/DependencyInjection/Compiler/AutowirePass.php

```diff
@@ -456,19 +463,21 @@ private function createAutowiredDefinition(\ReflectionClass $typeHint)
             return;
         }
 
-        $currentId = $this->currentId;
-        $this->currentId = $argumentId = sprintf('autowired.%s', $typeHint->name);
-
-        $argumentDefinition = $this->container->register($argumentId, $typeHint->name);
+        $argumentDefinition = $this->container->register($typeHint->name, $typeHint->name);
         $argumentDefinition->setPublic(false);
         $argumentDefinition->setAutowired(true);
+        $argumentDefinition->addTag('autoregistered');
 
-        $this->populateAvailableType($argumentId, $argumentDefinition);
+        if (null !== $this->types) {
+            $this->populateAvailableType($typeHint->name, $argumentDefinition);
+        }
 
-        $this->processValue($argumentDefinition);
+        $currentId = $this->currentId;
+        $this->currentId = $typeHint->name;
+        $this->processValue($argumentDefinition, true);
```

Comment:
sameh here: adding "true" has no practical side effect, but is just more correct at the semantical level